matchzoo/contrib/models/esim.py

```diff
@@ -0,0 +1,256 @@
+
+        :param texts: np.array with shape [B, T, H]
+        :param lengths: np.array with shape [B, T, ], where 1 means valid, 0 means pad
+        """
+        mask = self._expand_dim(mask, axis=2)
+        new_texts = Multiply()([texts, mask])
+
+        text_max = Lambda(
+            lambda x: K.max(x, axis=1),
+        )(new_texts)
+
+        return text_max
+
+
+    def _avg(self, texts, mask):
+        """
+        Compute the mean of each text according to their real length
+
+        :param texts: np.array with shape [B, T, H]
+        :param lengths: np.array with shape [B, T, ], where 1 means valid, 0 means pad
+        """
+        mask = self._expand_dim(mask, axis=2)
+        new_texts = Multiply()([texts, mask])
+
+        # timestep-wise division, exclude the PAD number when calc avg
+        text_avg = keras.layers.Lambda(
+            lambda text_mask: K.sum(text_mask[0], axis=1) / K.sum(text_mask[1], axis=1),
+        )([new_texts, mask])
+
+        return text_avg
+
+    def build(self):
+        """Build model."""
+
+        # parameters
+        lstm_dim = self._params['lstm_dim']
+        dropout_rate = self._params['dropout_rate']
+
+        # layers
+        create_mask = Lambda(lambda x: K.cast(K.not_equal(x, self._params['mask_value']), K.floatx()))
+        embedding = self._make_embedding_layer()
+        lstm_compare = self._make_BiLSTM_layer(lstm_dim)
+        lstm_compose = self._make_BiLSTM_layer(lstm_dim)
+        dense_compare = Dense(units=lstm_dim, activation='relu', use_bias=True)
+        dropout = Dropout(dropout_rate)
+
+
+        # model
+        a, b = self._make_inputs()      # [B, T_a], [B, T_b]
+        a_mask = create_mask(a)         # [B, T_a]
+        b_mask = create_mask(b)         # [B, T_b]
+
+        ########################
+        # encoding
+        ########################
+        a_emb = dropout(embedding(a))   # [B, T_a, E_dim]
+        b_emb = dropout(embedding(b))   # [B, T_b, E_dim]
+
+        a_ = lstm_compare(a_emb)          # [B, T_a, H*2]
+        b_ = lstm_compare(b_emb)          # [B, T_b, H*2]
+
+
+        ########################
+        # local inference
+        ########################
+        # similarity matrix
+        e = Dot(axes=-1)([a_, b_])  # [B, T_a, T_b]
+        _ab_mask = Multiply()([self._expand_dim(a_mask, axis=2),    # [B, T_a, 1]
+                               self._expand_dim(b_mask, axis=1)])   # [B, 1, T_b]
+        # _ab_mask: [B, T_a, T_b]
+
+        pm = Permute((2, 1))
+        mask_layer = self._make_atten_mask_layer()
+        softmax_layer = Softmax(axis=-1)
+
+        e_a = softmax_layer(mask_layer([e, _ab_mask]))          # [B, T_a, T_b]
+        e_b = softmax_layer(mask_layer([pm(e), pm(_ab_mask)]))  # [B, T_b, T_a]
+
+        # alignment (a_t = a~, b_t = b~ )
+        a_t = Dot(axes=(2, 1))([e_a, b_])   # [B, T_a, H*2]
+        b_t = Dot(axes=(2, 1))([e_b, a_])   # [B, T_b, H*2]
+
```

Comment:
Oh that's actually a good point...I though the lstm would mask internally yet right, since the input it not zero the lstm might not be able to recognize them. Will fix it!

Reply:
The mask are added in the new version XD